core/trino-main/src/main/java/io/trino/sql/gen/OrCodeGenerator.java

```diff
@@ -48,58 +49,41 @@ public BytecodeNode generateExpression(BytecodeGeneratorContext generator)
                 .comment("OR")
                 .setDescription("OR");
 
-        block.append(generator.generate(left));
-
-        IfStatement ifLeftIsNull = new IfStatement("if left wasNull...")
-                .condition(wasNull);
+        block.push(false); // keep track of whether we've seen a null so far
 
         LabelNode end = new LabelNode("end");
-        ifLeftIsNull.ifTrue(new BytecodeBlock()
-                .comment("clear the null flag, pop left value off stack, and push left null flag on the stack (true)")
+        LabelNode returnTrue = new LabelNode("returnTrue");
+        for (int i = 0; i < terms.size(); i++) {
+            RowExpression term = terms.get(i);
+            block.append(generator.generate(term));
+
+            IfStatement ifWasNull = new IfStatement(format("if term %s wasNull...", i))
+                    .condition(wasNull);
+
+            ifWasNull.ifTrue()
+                    .comment("clear the null flag, pop residual value off stack, and push was null flag on the stack (true)")
+                    .pop(term.getType().getJavaType()) // discard residual value
+                    .pop(boolean.class) // discard the previous "we've seen a null flag"
+                    .push(true);
+
+            ifWasNull.ifFalse()
+                    .comment("if term is false, short circuit and return false")
```

Comment:
"if term is false, short circuit and return false" -> "if term is true, short circuit and return true"